### PR TITLE
task done

### PR DESCRIPTION
**Pull request by yutsevich**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        IntStream.range(0, orders.size())
+                .filter(i -> orders.get(i).getId().equals(order.getId()))
+                .forEach(i -> orders.set(i, order));
+        return order;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.orders
+                .removeIf(o -> o.getId().equals(id));
```
Comment (andriylagun) [inline src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]: remove this separation before removeIf

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        IntStream.range(0, shoppingCarts.size())
+                .filter(i -> shoppingCarts.get(i).getId().equals(shoppingCart.getId()))
+                .forEach(i -> shoppingCarts.set(i, shoppingCart));
+        return shoppingCart;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.shoppingCarts
+                .removeIf(sc -> sc.getId().equals(id));
```
Comment (andriylagun) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartDaoImpl.java]: same

File: src/main/java/com/internet/shop/dao/impl/UserDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                .filter(i -> users.get(i).getId().equals(user.getId()))
+                .forEach(i -> users.set(i, user));
+        return user;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.users
+                .removeIf(u -> u.getId().equals(id));
+    }
```
Comment (andriylagun) [inline src/main/java/com/internet/shop/dao/impl/UserDaoImpl.java]: same

File: src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -1,7 +1,56 @@
...
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        List<Product> products = List.copyOf(shoppingCart.getProducts());
+        Long userId = shoppingCart.getUserId();
+        Order order = new Order(userId);
+        order.setProducts(products);
+        Storage.addOrder(order);
+        shoppingCartDao.update(new ShoppingCart(shoppingCart.getUserId()));
+        return order;
+    }
```
Comment (andriylagun) [inline src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java]: I think will be better use shoppingCart.getProducts().clear();

File: src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java [hunk 5]
```diff
@@ -1,7 +1,56 @@
...
+        return shoppingCartDao.get(shoppingCart.getId()).get()
+                .getProducts()
+                .removeIf(p -> p.getId().equals(product.getId()));
+    }
+
+    @Override
+    public void clear(ShoppingCart shoppingCart) {
+        Storage.shoppingCarts.stream()
+                .filter(sc -> sc.getId().equals(shoppingCart.getId()))
+                .forEach(sc -> sc.setProducts(new ArrayList<Product>()));
```
Comment (andriylagun) [inline src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java]: same here, you can use shoppingCart.getProducts().clear()

File: src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -1,7 +1,56 @@
...
+    @Inject
+    private ShoppingCartDao shoppingCartDao;
+
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        List<Product> products = List.copyOf(shoppingCart.getProducts());
+        Long userId = shoppingCart.getUserId();
+        Order order = new Order(userId);
+        order.setProducts(products);
+        Storage.addOrder(order);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java]: - we can't use Storage at the service layer
- let's call method clear() on the shoppingCart

File: src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -1,7 +1,56 @@
...
+        shoppingCartDao.update(new ShoppingCart(shoppingCart.getUserId()));
+        return order;
+    }
+
+    @Override
+    public List<Order> getUserOrders(Long userId) {
+        return Storage.orders.stream()
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java]: let's move this method to the dao layer

File: src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -1,7 +1,56 @@
...
+    public ShoppingCart create(ShoppingCart shoppingCart) {
+        return shoppingCartDao.create(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCart addProduct(ShoppingCart shoppingCart, Product product) {
+        shoppingCartDao.get(shoppingCart.getId()).get()
+                .getProducts()
+                .add(product);
+        return shoppingCartDao.get(shoppingCart.getId()).get();
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java]:
```suggestion
        shoppingCart.getProducts().add(product);
        return shoppingCartDao.update(shoppingCart);
```
подібно зробити в методі deleteProduct і clear()

File: src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -1,7 +1,56 @@
...
+
+    @Override
+    public void clear(ShoppingCart shoppingCart) {
+        Storage.shoppingCarts.stream()
+                .filter(sc -> sc.getId().equals(shoppingCart.getId()))
+                .forEach(sc -> sc.setProducts(new ArrayList<Product>()));
+    }
+
+    @Override
+    public ShoppingCart getByUserId(Long userId) {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java:41]: let's move this method to the dao layer

File: src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -22,31 +20,26 @@ public ShoppingCart create(ShoppingCart shoppingCart) {
...
     }
 
     @Override
     public boolean deleteProduct(ShoppingCart shoppingCart, Product product) {
-        return shoppingCartDao.get(shoppingCart.getId()).get()
-                .getProducts()
-                .removeIf(p -> p.getId().equals(product.getId()));
+        shoppingCart.getProducts().remove(product);
+        shoppingCartDao.update(shoppingCart);
+        return true;
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/servise/impl/ShoppingCartServiceImpl.java]:
```suggestion
        boolean result = shoppingCart.getProducts().remove(product);
        shoppingCartDao.update(shoppingCart);
        return result;
```

File: src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -1,7 +1,52 @@
...
+    private ShoppingCartDao shoppingCartDao;
+
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        List<Product> products = List.copyOf(shoppingCart.getProducts());
+        Long userId = shoppingCart.getUserId();
+        Order order = new Order(userId);
+        order.setProducts(products);
+        orderDao.create(order);
+        shoppingCart.getProducts().clear();
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/servise/impl/OrderServiceImpl.java]:
```suggestion
        shoppingCartService.clear(shoppingCart);
```

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**